### PR TITLE
esri-no-headers: fix the isDefault check to support templates

### PR DESCRIPTION
**Pull request by dkalowsk**
See commits for comments on them.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/esri/NoImplementationInHeadersCheck.cpp [hunk 1]
```diff
@@ -174,13 +159,23 @@ void NoImplementationInHeadersCheck::check(
...
+        // Issue: this allows ALL destructors to be declared in a header.  Currently
+        // there is no known method to fix this.
+        return;
+      }
       diag(FD->getLocation(),
-           "rule of five function %0 defaulted in a header file; "
-           "runtimecore prohibits defaults in headers. Please move them to the "
-           "source file")
+           "special member function %0 defaulted in a header file; "
+           "runtimecore prohibits defaults in headers and should be the source "
```
[the remark this answers is not in the document]
Comment (dkalowsk) [reply clang-tools-extra/clang-tidy/esri/NoImplementationInHeadersCheck.cpp]: Trying to insert your text, I'm failing.  If you've got a full sentence I'm all for it.  This was the best I thought of.

[the remark this answers is not in the document]
Comment (chri7325) [reply clang-tools-extra/clang-tidy/esri/NoImplementationInHeadersCheck.cpp]: runtimecore prohibits defaults in headers, with the exception of destructors, and should be defaulted in the source file

[the remark this answers is not in the document]
Comment (dkalowsk) [reply clang-tools-extra/clang-tidy/esri/NoImplementationInHeadersCheck.cpp]: Done

File: clang-tools-extra/clang-tidy/esri/NoImplementationInHeadersCheck.cpp [hunk 2]
```diff
@@ -174,13 +159,23 @@ void NoImplementationInHeadersCheck::check(
...
+        // Issue: this allows ALL destructors to be declared in a header.  Currently
+        // there is no known method to fix this.
+        return;
+      }
       diag(FD->getLocation(),
-           "rule of five function %0 defaulted in a header file; "
-           "runtimecore prohibits defaults in headers. Please move them to the "
-           "source file")
+           "special member function %0 defaulted in a header file; "
+           "runtimecore prohibits defaults in headers, with the exception "
```
[the remark this answers is not in the document]
Comment (dkalowsk) [reply clang-tools-extra/clang-tidy/esri/NoImplementationInHeadersCheck.cpp]: That is pretty much all over the module as it is today.  I'd prefer to leave it in so we know it's a runtime specific check.

**Pull request merged**